pkg/proxy/util/nodeport_addresses.go

```diff
@@ -24,23 +24,57 @@ import (
 	netutils "k8s.io/utils/net"
 )
 
-// GetNodeAddresses return all matched node IP addresses based on given cidr slice.
-// Some callers, e.g. IPVS proxier, need concrete IPs, not ranges, which is why this exists.
+// NodePortAddresses is used to handle the --nodeport-addresses flag
+type NodePortAddresses struct {
+	cidrStrings []string
+
+	cidrs                []*net.IPNet
+	containsIPv4Loopback bool
+}
+
+// RFC 5735 127.0.0.0/8 - This block is assigned for use as the Internet host loopback address
+var ipv4LoopbackStart = net.IPv4(127, 0, 0, 0)
+
+// NewNodePortAddresses takes the `--nodeport-addresses` value (which is assumed to
+// contain only valid CIDRs) and returns a NodePortAddresses object. If cidrStrings is
+// empty, this is treated as `["0.0.0.0/0", "::/0"]`.
+func NewNodePortAddresses(cidrStrings []string) *NodePortAddresses {
+	if len(cidrStrings) == 0 {
+		cidrStrings = []string{IPv4ZeroCIDR, IPv6ZeroCIDR}
+	}
+
+	npa := &NodePortAddresses{
+		cidrStrings: cidrStrings,
+	}
+
+	for _, str := range npa.cidrStrings {
+		_, cidr, _ := netutils.ParseCIDRSloppy(str)
+		npa.cidrs = append(npa.cidrs, cidr)
+
+		if netutils.IsIPv4CIDR(cidr) {
+			if cidr.IP.IsLoopback() || cidr.Contains(ipv4LoopbackStart) {
+				npa.containsIPv4Loopback = true
+			}
+		}
+	}
+
+	return npa
+}
+
+func (npa *NodePortAddresses) String() string {
+	return fmt.Sprintf("%v", npa.cidrStrings)
+}
+
+// GetNodeAddresses return all matched node IP addresses for npa's CIDRs.
+// If npa's CIDRs include "0.0.0.0/0" and/or "::/0", then those values will be returned
+// verbatim in the response and no actual IPs of that family will be returned.
+// If no matching IPs are found, GetNodeAddresses will return an error.
 // NetworkInterfacer is injected for test purpose.
-// We expect the cidrs passed in is already validated.
-// Given an empty input `[]`, it will return `0.0.0.0/0` and `::/0` directly.
-// If multiple cidrs is given, it will return the minimal IP sets, e.g. given input `[1.2.0.0/16, 0.0.0.0/0]`, it will
-// only return `0.0.0.0/0`.
-// NOTE: GetNodeAddresses only accepts CIDRs, if you want concrete IPs, e.g. 1.2.3.4, then the input should be 1.2.3.4/32.
-func GetNodeAddresses(cidrs []string, nw NetworkInterfacer) (sets.String, error) {
+func (npa *NodePortAddresses) GetNodeAddresses(nw NetworkInterfacer) (sets.String, error) {
 	uniqueAddressList := sets.NewString()
-	if len(cidrs) == 0 {
-		uniqueAddressList.Insert(IPv4ZeroCIDR)
-		uniqueAddressList.Insert(IPv6ZeroCIDR)
-		return uniqueAddressList, nil
-	}
+
 	// First round of iteration to pick out `0.0.0.0/0` or `::/0` for the sake of excluding non-zero IPs.
-	for _, cidr := range cidrs {
+	for _, cidr := range npa.cidrStrings {
```

Comment:
... IIUIC you still have process a slice of strings, the difference is the Proxies now has a struct as field instead of a slice

Reply:
I apologize but I can't see the optimization , AFAIK the number of operations are still the same

Reply:
oh, oops, yeah, that particular loop [goes away later](https://github.com/danwinship/kubernetes/commit/6e96baafd7fee03e50e8ae246f6a0f445bbdfe5e#diff-f1a66079e980f4783267715c848dd2e1c235032d49d40c590c70fc4439cd0643)  :slightly_smiling_face:

Reply:
but this will be always be a constant, and usually very small.

The thing I can see as an improvement is to replace the list of the host interfaces in each iteration, I think that @trozet implemented something in ovn-kubernetes to watch the interfaces on the host instead of polling them

Reply:
If you want me to remove the word "optimize" from the commit message, I can.

The real fix this is building toward is cleaning up the ugly API (where `GetNodeAddresses()` returns a special magic value to indicate "matches all IPs" which the caller needs to handle specially). In the process of doing that, I noticed that we were re-parsing the list of CIDRs on every run through the loop, which seemed dumb, so I fixed it to not do that.

Reply:
oh, no, sorry, I was trying to understand the PR, I didn't want to nitpick.

I think that I know understand what you are trying to do based on the comments in the KEP for moving this to staging, to build a library and people can use GetNodeAddresses to configure NodePorts ...